lib/puppet/pops/loaders.rb

```diff
@@ -43,7 +44,10 @@ def initialize(environment, for_agent)
     #
     @puppet_system_loader = create_puppet_system_loader()
 
-    # 2. Environment loader - i.e. what is bound across the environment, may change for each setup
+    # 2. Cache loader - ie what puppet stores on disk via pluginsync.
+    @puppet_cache_loader = create_puppet_cache_loader()
```

Comment:
I'm not sure what the implications are for puppetserver here, particularly compile masters; do those machines ever use pluginsync to move files around?

Additionally, is this in the correct order? Being placed before the loader that loads the puppet internal functions allows for users to prioritize their own custom definitions before puppet's, but that seems to already be the case with environment loaders preceding the system_loader.

Reply:
In a multi-master scenario, the agent running on the compile master will pluginsync from the mom (master of masters), and some people rely on that to distribute server-side extensions like hiera backends. In Puppet 4 we changed puppetserver to exclude the agent's libdir, so the default is:

```hocon
    ruby-load-path: [/opt/puppetlabs/puppet/lib/ruby/vendor_ruby]
```

That broke some people (https://tickets.puppetlabs.com/browse/SERVER-571), and the current workaround is to add the agent's libdir back to `ruby-load-path`, with the understanding that it will break environment isolation (since any types/providers that the compile master's agent pluginsync will be visible across all environments when the compile master compiles catalogs for other agents).

Reply:
There's another work around, which is to schedule r10k deployments on each compile master when a control repo changes (diy code-manager). Then users may use a post deploy hook to generate types and retain env isolation.